File: app/src/main/java/eu/warble/pjapp/ui/main/ftp/FtpFragment.java
package eu.warble.pjapp.ui.main.ftp;


import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.design.widget.Snackbar;
import android.support.v4.app.FragmentActivity;
import android.support.v7.app.ActionBar;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.SearchView;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.LinearLayout;
import android.widget.ListView;

import com.jcraft.jsch.ChannelSftp;

import java.util.Vector;

import eu.warble.pjapp.R;
import eu.warble.pjapp.ui.base.BaseFragment;

public class FtpFragment extends BaseFragment<FtpPresenter> implements SearchView.OnQueryTextListener{

    private ListView listView;
    private LinearLayout loadingScreen;
    private SearchView searchView;
    private MenuItem searchItem;

    public static FtpFragment newInstance(){
        FtpFragment ftpFragment = new FtpFragment();
        ftpFragment.tag = "fragment_ftp";
        return ftpFragment;
    }

    @Override
    protected FtpPresenter createPresenter() {
        return new FtpPresenter(this);
    }

    @Override
    protected View initView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_ftp, container, false);
        setHasOptionsMenu(true);
        initToolbar();
        initViews(view);
        return view;
    }

    private void initViews(View view){
        loadingScreen = view.findViewById(R.id.loading_screen);
        listView = view.findViewById(R.id.ftpListView);
        listView.setAdapter(new FtpListAdapter(context));
        listView.setOnItemClickListener((parent, view1, position, id) -> {
            ChannelSftp.LsEntry item = (ChannelSftp.LsEntry) listView.getAdapter().getItem(position);
            if(item.getAttrs().isDir()){
                presenter.loadDirectory(item.getFilename());
                if(!searchView.isIconified()){
                    searchItem.collapseActionView();
                }
            }else {
                presenter.downloadFile(item.getFilename());
            }
        });
    }

    public void showLoadingScreen(){
        loadingScreen.setVisibility(View.VISIBLE);
    }

    public void hideLoadingScreen(){
        loadingScreen.setVisibility(View.GONE);
    }

    public void updateList(Vector<ChannelSftp.LsEntry> newData){
        FtpListAdapter adapter = (FtpListAdapter) listView.getAdapter();
        adapter.updateList(newData);
    }

    public void showMessage(String message){
        FragmentActivity activity = getActivity();
        if (activity != null)
            Snackbar.make(activity.findViewById(android.R.id.content), message, Snackbar.LENGTH_LONG).show();
    }

    private void initToolbar(){
        AppCompatActivity activity = ((AppCompatActivity)getActivity());
        ActionBar bar = activity.getSupportActionBar();
        if (bar != null) {
            bar.setTitle(R.string.pjatk);
            bar.setElevation(14f);
        }
    }

    @Override
    public void onCreateOptionsMenu(Menu menu, MenuInflater inflater) {
        inflater.inflate(R.menu.toolbar, menu);
        searchItem = menu.findItem(R.id.action_search);
        searchItem.setEnabled(true);
        searchItem.setVisible(true);
        searchView = (SearchView) searchItem.getActionView();
        searchView.setOnQueryTextListener(this);
    }

    @Override
    public boolean onQueryTextSubmit(String query) {
        return true;
    }

    @Override
    public boolean onQueryTextChange(String newText) {
        if (loadingScreen.getVisibility() == View.GONE)
            ((FtpListAdapter)listView.getAdapter()).getFilter().filter(newText);
        return true;
    }
}

File: app/src/main/java/eu/warble/pjappkotlin/mvp/ftp/FtpFragment.kt
package eu.warble.pjappkotlin.mvp.ftp

import android.Manifest
import android.annotation.SuppressLint
import android.content.Context
import android.os.Bundle
import android.support.v7.app.AlertDialog
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView
import android.support.v7.widget.SearchView
import android.view.LayoutInflater
import android.view.Menu
import android.view.MenuInflater
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup
import eu.warble.getter.model.GetterFile
import eu.warble.pjappkotlin.R
import eu.warble.pjappkotlin.mvp.ApplicationNavigator
import eu.warble.pjappkotlin.mvp.BaseActivity
import eu.warble.pjappkotlin.mvp.BaseFragment
import eu.warble.pjappkotlin.utils.Injection
import kotlinx.android.synthetic.main.fragment_ftp.view.directories_list
import kotlinx.android.synthetic.main.fragment_ftp.view.loading_screen
import permissions.dispatcher.NeedsPermission
import permissions.dispatcher.OnShowRationale
import permissions.dispatcher.PermissionRequest
import android.widget.Toast
import permissions.dispatcher.OnPermissionDenied
import permissions.dispatcher.RuntimePermissions

@RuntimePermissions
class FtpFragment : BaseFragment(), FtpContact.View, SearchView.OnQueryTextListener {

    override val TAG: String = "fragment_ftp"
    override lateinit var presenter: FtpContact.Presenter
    override val applicationNavigator: ApplicationNavigator by lazy {
        ApplicationNavigator(activity as BaseActivity)
    }
    private lateinit var loadingScreen: View
    private lateinit var directoriesList: RecyclerView
    private lateinit var searchView: SearchView
    private lateinit var searchItem: MenuItem

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        val view = inflater.inflate(R.layout.fragment_ftp, container, false)
        setHasOptionsMenu(true)
        initViews(view)
        initPresenter()
        presenter.start()
        return view
    }

    private fun initViews(view: View) {
        loadingScreen = view.loading_screen
        directoriesList = view.directories_list
        directoriesList.layoutManager = LinearLayoutManager(directoriesList.context)
        directoriesList.adapter = FtpListAdapter(
                emptyList(),
                onItemClick = { file: GetterFile ->
                    if (file.isDirectory()) {
                        presenter.loadDirectory(file.path)
                        if (!searchView.isIconified)
                            searchItem.collapseActionView()
                    } else {
                        downloadFileWithPermissionCheck(file)
                    }
                }
        )
    }

    override fun showLoadingScreen(show: Boolean) {
        loadingScreen.visibility = if (show) View.VISIBLE else View.GONE
    }

    override fun updateDirectoriesList(newData: ArrayList<GetterFile>) {
        (directoriesList.adapter as FtpListAdapter).updateList(newData)
    }

    private fun initPresenter() {
        presenter = FtpPresenter(
                Injection.provideCredentials(mContext as Context),
                this,
                mContext as Context
        )
    }

    @NeedsPermission(Manifest.permission.WRITE_EXTERNAL_STORAGE, Manifest.permission.READ_EXTERNAL_STORAGE)
    fun downloadFile(file: GetterFile) {
        presenter.downloadFile(file)
        showMessage(getString(R.string.download_start))
    }

    @OnShowRationale(Manifest.permission.WRITE_EXTERNAL_STORAGE, Manifest.permission.READ_EXTERNAL_STORAGE)
    fun showRationaleForStorage(request: PermissionRequest) {
        AlertDialog.Builder(mContext as Context)
                .setMessage(R.string.permission_rationale_storage_message)
                .setPositiveButton(R.string.button_allow, { _, _ -> request.proceed() })
                .setNegativeButton(R.string.button_deny, { _, _ -> request.cancel() })
                .setCancelable(false)
                .show()
    }

    @OnPermissionDenied(Manifest.permission.WRITE_EXTERNAL_STORAGE, Manifest.permission.READ_EXTERNAL_STORAGE)
    fun showDeniedForStorage() {
        Toast.makeText(activity, R.string.permissions_declined, Toast.LENGTH_SHORT).show()
    }

    @SuppressLint("NeedOnRequestPermissionsResult")
    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<out String>, grantResults: IntArray) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
        onRequestPermissionsResult(requestCode, grantResults)
    }

    override fun onCreateOptionsMenu(menu: Menu, inflater: MenuInflater) {
        menu.clear()
        inflater.inflate(R.menu.toolbar, menu)
        searchItem = menu.findItem(R.id.action_search).apply {
            isEnabled = true
            isVisible = true
        }
        searchView = searchItem.actionView as SearchView
        searchView.setOnQueryTextListener(this)
    }

    override fun onQueryTextChange(newText: String): Boolean {
        if (loadingScreen.visibility == View.GONE)
            (directoriesList.adapter as FtpListAdapter).filter.filter(newText)
        return true
    }

    override fun onQueryTextSubmit(query: String): Boolean {
        return true
    }

    override fun onBack(): Boolean {
        return presenter.onBack()
    }

    companion object {
        fun newInstance() = FtpFragment()
    }
}